src/libp2p_multistream_client.erl

```diff
@@ -41,11 +45,19 @@ ls(Connection) ->
         {error, Error} -> {error, Error}
     end.
 
--spec handshake(libp2p_connection:connection()) -> ok | {error, term()}.
+-spec handshake(libp2p_connection:connection()) -> ok | server_switch | {error, term()}.
 handshake(Connection) ->
     Id = libp2p_multistream:protocol_id(),
-    case libp2p_multistream:read(Connection) of
-        Id -> libp2p_multistream:write(Connection, Id);
+    case libp2p_multistream:read(Connection, rand:uniform(20000) + 15000) of
+        Id ->
+            ok = libp2p_multistream:write(Connection, Id);
+        {error, timeout} ->
+            ok = libp2p_multistream:write(Connection, Id),
+            case libp2p_multistream:read(Connection) of
+                Id ->
+                    server_switch;
```

Comment:
Could you leave some commentary here our forgetful future selves know why the hell we did all this magic?